sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroReadSchemaTransformProvider.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io;
+
+import com.google.auto.value.AutoValue;
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.List;
+import org.apache.beam.sdk.schemas.AutoValueSchema;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.schemas.annotations.DefaultSchema;
+import org.apache.beam.sdk.schemas.transforms.Convert;
+import org.apache.beam.sdk.schemas.transforms.SchemaTransform;
+import org.apache.beam.sdk.schemas.transforms.TypedSchemaTransformProvider;
+import org.apache.beam.sdk.schemas.utils.AvroUtils;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.values.PCollectionRowTuple;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Strings;
+
+public class AvroReadSchemaTransformProvider
+    extends TypedSchemaTransformProvider<
+        AvroReadSchemaTransformProvider.AvroReadSchemaTransformConfiguration> {
+  @Override
+  public String identifier() {
+    return "beam:schematransform:org.apache.beam:avro_read:v1";
+  }
+
+  @Override
+  public List<String> inputCollectionNames() {
+    return Collections.emptyList();
+  }
+
+  @Override
+  public List<String> outputCollectionNames() {
+    return Collections.singletonList("output");
+  }
+
+  @Override
+  protected Class<AvroReadSchemaTransformConfiguration> configurationClass() {
+    return AvroReadSchemaTransformConfiguration.class;
+  }
+
+  @Override
+  protected SchemaTransform from(AvroReadSchemaTransformConfiguration configuration) {
+    return new AvroReadSchemaTransform(configuration);
+  }
+
+  static class AvroReadSchemaTransform implements SchemaTransform {
+
+    AvroReadSchemaTransformConfiguration config;
+
+    AvroReadSchemaTransform(AvroReadSchemaTransformConfiguration config) {
+      this.config = config;
+    }
+
+    @Override
+    public PTransform<PCollectionRowTuple, PCollectionRowTuple> buildTransform() {
+      return new PTransform<PCollectionRowTuple, PCollectionRowTuple>() {
+        @Override
+        public PCollectionRowTuple expand(PCollectionRowTuple input) {
```

Comment:
My understanding was that we usually split the PTransform this builds into its own class, called a PCollectionRowTupleTransform, which would allow the logic to be broken down a little more. As it stands, this implementation, with the nested return statements, is a bit arcane to me. I can figure out what it is doing by referencing other SchemaTransformProviders but it isn't as obvious as it could be, I think.

Reply:
Moved expansion to its own class for clarity